File: client/src/main/java/communication/Exchanger.java
package communication;

import collection.CollectionChangelist;
import collection.CollectionWrapper;
import commands.CommandNameToInfo;
import commands.ExecutionPayload;
import commands.ExecutionResult;
import exceptions.InvalidCredentialsException;
import model.data.Model;
import user.User;

import java.io.IOException;

public interface Exchanger {
  void requestFullCollection();

  void requestCollectionChanges(Long currentVersion);

  void requestCommandExecution(ExecutionPayload payload);

  void requestAccessibleCommandsInfo();

  <T extends Model> CollectionChangelist<T> receiveCollectionChanges() throws IOException, InvalidCredentialsException;

  void requestLogin(User user);

  void requestRegister(User user);

  <T extends Model> CollectionWrapper<T> receiveFullCollection() throws IOException, InvalidCredentialsException;

  ExecutionResult receiveExecutionResult() throws IOException, InvalidCredentialsException;

  CommandNameToInfo receiveAccessibleCommandsInfo() throws IOException, InvalidCredentialsException;

  void validateLogin() throws InvalidCredentialsException, IOException;

  void validateRegister() throws InvalidCredentialsException, IOException;
}

File: client-ui/common/src/commonMain/kotlin/common/context/Exchanger.kt
package common.context

import collection.CollectionChangelist
import collection.CollectionWrapper
import commands.CommandNameToInfo
import commands.ExecutionPayload
import commands.ExecutionResult
import common.exceptions.InvalidCredentialsException
import model.data.Model
import user.User

interface Exchanger {
    fun <T : Model> requestFullCollection(): CollectionWrapper<T>
    fun <T : Model> requestCollectionChanges(currentVersion: Long): CollectionChangelist<T>
    fun requestCommandExecution(payload: ExecutionPayload): ExecutionResult
    fun requestAccessibleCommandsInfo(): CommandNameToInfo

    fun requestLogin(user: User)
    fun requestRegister(user: User)
}